### PR TITLE
Rename / restructure input / output setup in filebeat

### PR DESCRIPTION
**Pull request by ruflin**
Naming convention:
- Everything using a channel to send data is called input / output with a Send method
- If only confirmation of events happens, it is called a Logger with a log method

More cleanup is needed in future PRs.

This replaces https://github.com/elastic/beats/pull/2511

### CODE REVIEW

File: filebeat/beater/filebeat.go [hunk 1]
```diff
@@ -98,9 +99,9 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	publisher.Start()
 	// Stopping publisher (might potentially drop items)
 	defer func() {
-		// Closes first the registrar logger to make sure not more events arrive at the registrar
-		// registrarChannel must be closed first to potentially unblock (pretty unlikely) the publisher
-		registrarChannel.Close()
+		// Closes first the publisher outpt to make sure not more events are sent to the registry
```
Comment (tsg) [inline filebeat/beater/filebeat.go]: tiny typo: "outpt".

File: filebeat/publisher/publisher.go [hunk 2]
```diff
@@ -19,23 +19,22 @@ type LogPublisher interface {
 	Stop()
 }
 
-// SuccessLogger is used to report successfully published events.
-type SuccessLogger interface {
-
+// Logger is used to report successfully published events.
+type Output interface {
```
Comment (tsg) [inline filebeat/publisher/publisher.go]: Type is Output, the comment says "Logger".

**Comment by tsg** [comment]
TBH these naming conventions make the code more abstract to me. But I defer to you and @urso on what's best here.

**Comment by ruflin** [comment]
@tsg I actually hoped to reach the opposite especially for the variable names. They should now say much more what they are then it was before. For the naming conventions, the problem before was that there was Outlet / Output but not really a difference and the naming was not consistent. Any specific parts you stumbled over which are now too abstract?

File: filebeat/beater/filebeat.go (hunk 1 above)
Comment (ruflin) [inline filebeat/beater/filebeat.go]: fixed

File: filebeat/publisher/publisher.go (hunk 2 above)
Comment (ruflin) [inline filebeat/publisher/publisher.go]: done

File: filebeat/beater/channels.go [hunk 3]
```diff
@@ -5,46 +5,43 @@ import (
...
-type finishedLogger struct {
+type logger struct {
 	wg *sync.WaitGroup
 }
 
-func newSpoolerOutlet(
+func newProspectorOutput(
 	done <-chan struct{},
-	s *spooler.Spooler,
+	input chan *input.Event,
```
Comment (urso) [inline filebeat/beater/channels.go]: while this one uses the spooler channel  only, idea of spoolerOutlet was to wrap/hide the spooler and the fact we're dealing with a channel. I'd even prefer in future calling a spooler.OnEvent directly, hiding the fact we're dealing with an asynchronous worker.

e.g. have a synchronous spooler worker with 'flush' upon full + potential async wrapper:

```
type spooler struct {}
type asyncSpooler struct {}

func (s *spooler) OnEvent()
func (s *spooler) Flush()

func newAsyncSpooler(s *spooler) *asyncSpooler
func (s *asyncSpooler) OnEvent()
func (s *asyncSpooler) Stop()

```

Both types support OnEvent, so no one needs to know about channels here.

File: filebeat/beater/channels.go [hunk 4]
```diff
@@ -5,46 +5,43 @@ import (
...
 
-type spoolerOutlet struct {
-	wg      *sync.WaitGroup
-	done    <-chan struct{}
-	spooler *spooler.Spooler
-
+type prospectorOutput struct {
+	wg     *sync.WaitGroup
+	done   <-chan struct{}
+	input  chan *input.Event
```
Comment (urso) [inline filebeat/beater/channels.go]: see comment on `newProspectorOutput`

File: filebeat/beater/channels.go [hunk 5]
```diff
@@ -5,46 +5,43 @@ import (
...
 }
 
-type registrarLogger struct {
+type publisherOutput struct {
 	done chan struct{}
 	ch   chan<- []*input.Event
 }
 
-type finishedLogger struct {
+type logger struct {
```
Comment (urso) [inline filebeat/beater/channels.go]: while `finishedLogger` is a bad naming, I wouldn't call it `logger` either, as logger somehow implies this is about logging a message to file.

File: filebeat/beater/filebeat.go [hunk 6]
```diff
@@ -51,32 +51,33 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
...
+	publisherOutput := newPublisherOutput(registrar.GetInput())
 
-	// Channel from spooler to harvester
-	publisherChan := newPublisherChannel()
+	// Channel from spooler to publisher
+	spoolerOutput := newSpoolerOutput()
 
 	// Publishes event to output
-	publisher := publisher.New(config.PublishAsync, publisherChan.ch, registrarChannel, b.Publisher)
+	publisher := publisher.New(config.PublishAsync, spoolerOutput.ch, publisherOutput, b.Publisher)
```
Comment (urso) [inline filebeat/beater/filebeat.go]: here I'd prefer having something like `registrarSink`... or directly have a call like:

```
publisher.New(config.PublishAsync, pubEventSource(spoolerOutput), onFinished(registrar), b.Publisher)
```

to make more clear how events flow through the system.

File: filebeat/beater/filebeat.go [hunk 7]
```diff
@@ -51,32 +51,33 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
...
 
 	// Init and Start spooler: Harvesters dump events into the spooler.
-	spooler, err := spooler.New(config, publisherChan)
+	spooler, err := spooler.New(config, spoolerOutput)
```
Comment (urso) [inline filebeat/beater/filebeat.go]: idea of using `publisherChan` (while naming is bad), was to make more clear which component the spooler outputs to.

File: filebeat/prospector/prospector.go [hunk 8]
```diff
@@ -38,15 +38,15 @@ type Prospectorer interface {
 	Run()
 }
 
-type Outlet interface {
-	OnEvent(event *input.Event) bool
+type Output interface {
+	Send(event *input.Event) bool
 }
```
Comment (urso) [inline filebeat/prospector/prospector.go]: this is not necessarily `Output`, but more like an `Eventer` with `OnEvent` method?

**Comment by ruflin** [comment]
After a discussion with @urso we decided to close this PR and instead open several smaller PR's that hide the channels from `filebeat.go` so no need anymore for the naming.

**Pull request closed without merging**